scheduler: KEP-5007 add integration tests on DeviceBindingConditions

ttsuuubasa opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

/kind feature

#### What this PR does / why we need it:

This PR adds integration tests for device preparation failure and binding condition timeout recovery.
- Simulates device preparation failure via device taints or removal, verifying that the scheduler reschedules the pod to a healthy node.
- Adds tests that validate scheduler behavior when device BindingConditions never become ready and when a new device without binding conditions becomes available.
- Exercises the full KEP-5007 “Device Binding Conditions” flow, including PreBind timeout handling and reallocation after failure.

**_Test scenarios added:_**

1. **DeviceBindingFailureConditionsReschedule**

    - The pod's ResourceClaim is first allocated to a device that has BindingConditions set.
    - Device preparation failure is simulated before setting BindingFailureConditions, using two approaches:
      - By tainting the device to prevent scheduling (NoSchedule taint) - if DRADeviceTaints enabled.
      - By removing the device from the ResourceSlice, simulating its unavailability - if DRADeviceTaints disabled.
    - A device without BindingConditions is added on a different node.
    - BindingFailureConditions is set to the ResourceClaim's status conditions to trigger rescheduling.
  
    **_Expected Behavior_**
    The claim is successfully reallocated to a device on a different node.
    The BindingFailure condition is cleared after rescheduling.
    The pod associated with the claim is eventually scheduled without error.

2. **DeviceBindingConditionsTimeoutReached**

    - Start the scheduler with a short bindingTimeout: 6s.
    - Create a ResourceSlice containing a single device (e.g. dev-w-binding) that requires a binding condition (BindingConditions: ["attached"]) and declares BindingFailureConditions (e.g. ["failed"]).
    - Create a ResourceClaim + Pod; the claim allocates to that device in the pool worker-0-with-binding.
    - Expect the scheduler’s PreBind to wait until the binding timeout and then surface a user-visible “binding timeout” message in the Pod’s conditions, while the Pod remains unscheduled (no Spec.NodeName and no PodScheduled=True condition).
  
    **_Expected Behavior_**
    PreBind timeout is enforced when required device conditions never become true; the Pod reports a user-visible message containing “binding timeout”, and the Pod stays unscheduled in the timeout-enforced scenario.

3. **DeviceBindingConditionsTimeoutRecovery**

    - Start the scheduler with bindingTimeout: 5s.
    - Create an initial ResourceSlice on worker-0-with-binding with a single device that requires BindingConditions: ["attached"] and BindingFailureConditions: ["failed"].
    - Create a ResourceClaim + Pod; wait until the claim is allocated to this condition-gated device.
    - While the scheduler is still in the PreBind wait window, create a second ResourceSlice on worker-0-without-binding with a single device that has no BindingConditions (e.g. dev-no-binding).
    - Wait until the Pod reports a binding timeout in one of its conditions (message contains “binding timeout” / “timed out”).
    - Expect the claim to be reallocated to the new device in the worker-0-without-binding pool and the Pod to successfully schedule onto the node.
    - Finally, verify that the claim’s Status.Allocation points to the new device and that Status.ReservedFor includes the scheduled Pod.
  
    **_Expected Behavior_**
    After a timeout on a condition-gated device, adding an alternative device without binding conditions allows the claim to reallocate and the Pod to schedule.

#### Which issue(s) this PR is related to:
<!--
Please link relevant issues to help with tracking.

To automatically close the linked issue(s) when this PR is merged,
add the word "Fixes" before the issue number or link.
Do not use "Fixes" if the PR is of kind `failing-test` or `flake`.

Reference KEPs when applicable in addition to specific issues.

Examples:
Fixes #<issue number>
<issue link> (issue in a different repository)
KEP: https://github.com/kubernetes/enhancements/issues/<kep-issue-number>

If there is no associated issue, then write "N/A".
-->

Part of [KEP-5007: DRA Device Binding Conditions](https://github.com/kubernetes/enhancements/issues/5007)

#### Special notes for your reviewer:

@johnbelamaric @dom4ha 

This PR is one of the enhancements of test for Device Binding Conditions (KEP-5007) targeting v1.35, and it fulfills one of the graduation criteria described in the KEP (originally listed as a beta graduation criterion).

- Additional tests are in Testgrid and linked in KEP

To proactively meet the criteria for beta, we are considering merging this in the v1.35 alpha release.

@pohly 

I’ve created the test cases you previously suggested in the [comment](https://github.com/kubernetes/kubernetes/pull/130160/files#r2218808034).
Could you please take a look when you have a moment?

```
Given that it's not clear whether the code is correct: more test scenarios. This function can be used for simple on/off checks and be called multiple times. Advanced tests then run only when the feature is enabled and should cover:

- some, but not all devices using binding conditions
- binding failure such that a different node must be used to finish scheduling:
  - when device taints are enabled, taint the device in the ResourceSlice before setting the binding failure condition
  - otherwise remove the failed device from the ResourceSlice before setting the binding failure condition
- binding times out while some other device gets added without binding conditions: the scheduler should recover by picking that other new device instead

The second case is different from the existing "set binding failure condition, remove it later". However, this hinges a bit on how we handle binding failure conditions. The current test might not work anymore as-is when trying to allocate something else instead immediately - I'm a bit fuzzy on that.
```

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/5007
```

ttsuuubasa commented on test/integration/dra/dra_test.go:
```diff
@@ -1686,6 +1661,156 @@ func testDeviceBindingConditions(tCtx ktesting.TContext, enabled bool) {
+				},
+			},
+		}
+		sliceWithoutBinding2 = createSlice(tCtx, sliceWithoutBinding2)
```
The variable `sliceWithoutBinding2` defined here is not used afterward, which causes the lint test to fail. I confirmed locally that replacing the return value with `_` allows the integration test to pass and the lint test to succeed. I plan to include this fix along with other comment updates.

ttsuuubasa commented:
@pohly 
To resolve the conflict with your PR #134973 , I made some modifications — such as moving the helper function to be planned for implementation to the newly created `helper_test.go`.

fj-naji commented:
/retest

tsj-30 commented:
@ttsuuubasa 
The structure closely matches the intention of the original test but isolates the “failure → retry → success” path nicely.
this looks good to me  👍

pohly commented on test/integration/dra/dra_test.go:
```diff
@@ -1593,4 +1562,155 @@ func testDeviceBindingConditions(tCtx ktesting.TContext, enabled bool) {
 	waitForPodScheduled(tCtx, namespace, pod.Name)
+
+	// When device preparation fails on a node and a BindingFailure is written, verify that rescheduling leads to a successful binding on a different node.
+	enableTaints := []bool{false, true}
```
Instead of testing two different scenarios in the same test, can you run the test twice, once with and once without taints?

Then when one test fails and the other works it's easier to debug.

ttsuuubasa replied:
Would it be a good idea to further call `tCtx.Run` inside `testDeviceBindingConditions` in order to split the tests?

pohly commented on test/integration/dra/dra_test.go:
```diff
@@ -1593,4 +1562,155 @@ func testDeviceBindingConditions(tCtx ktesting.TContext, enabled bool) {
+		}))), "third allocated claim to the device with binding conditions")
+
+		// Emulate a device preparation failure scenario by either tainting the device or removing it.
+		featuregatetesting.SetFeatureGateDuringTest(tCtx, utilfeature.DefaultFeatureGate, features.DRADeviceTaints, enableTaint)
```
Changing feature gates doesn't belong into the test. The boilerplate code sets feature gates and then calls tests which can run in that environment:

https://github.com/kubernetes/kubernetes/blob/38328aa3f82c97369f322f581ff98bc4e43e885a/test/integration/dra/dra_test.go#L122-L228

ttsuuubasa replied:
To achive this, what kind of test case names do you think should be added alongside `"all"` and `"GA"`?
For example:
- `"taint-disabled-bc-enabled"`: This means taint is disabled and BindingConditions is enabled.
- `"only-taint-disabled"`: This means only taint is disabled, and everything else is enabled.

Then, by adding an arg to the function like
  `func testDeviceBindingConditions(tCtx ktesting.TContext, bcEnabled bool, taintEnabled bool)`
we could run the case as `testDeviceBindingConditions(tCtx, true, false)`.

pohly replied:
Taints don't need to be disabled to test without them. Your function's parameter should be `useTaints bool`, not `taintsEnabled bool`. Then it can be invoked with `false` in the "All" cluster setup.

pohly commented on test/integration/dra/helpers_test.go:
```diff
@@ -188,3 +188,36 @@ func waitForNotFound[T any](tCtx ktesting.TContext, get func(context.Context, st
 }
+
+func waitForClaimAllocatedToDevice(tCtx ktesting.TContext, namespace, claimName string, timeout time.Duration) *resourceapi.ResourceClaim {
+	var latestClaim *resourceapi.ResourceClaim
```
```suggestion
	tCtx.Helper()
	var latestClaim *resourceapi.ResourceClaim
```

Test failures should be attributed to the caller location, not the helper function.

pohly commented on test/integration/dra/helpers_test.go:
```diff
@@ -188,3 +188,36 @@ func waitForNotFound[T any](tCtx ktesting.TContext, get func(context.Context, st
+			}
+		}
+		return latestClaim
+	}).WithTimeout(timeout).WithPolling(time.Second).Should(gomega.BeNil(), "claim should not have any condition")
```
This is going to be misleading if it times out while the claim has never had any conditions at all. 

Overall it would be better to have a generic `func waitForClaim(tCtx ktesting.TContext, namespace, claimName string, timeout time.Duration, match gtypes.GomegaMatcher, description ...any) *resourceapi.ResourceClaim` which gets called with different matches. For example:

```
func waitForClaimAllocatedToDevice(tCtx ktesting.TContext, namespace, claimName string, timeout time.Duration) *resourceapi.ResourceClaim {
    return waitForClaim(tCtx, namespace, claimName, timeout, gomega.HaveField("Status.Allocation", gomega.Not(gomega.BeNil())), "Claim should have been allocated.")
}
```

Then `waitDeviceConditionsBeNil` needs to be replaced with waiting for the condition and then waiting for the next. But I am not sure I fully grasp what this waits for. A rewrite should make it clearer.

ttsuuubasa committed:
scheduler: KEP-5007 move BindingConditions integration tests to separate file

To improve maintainability and reduce the size of dra_test.go,
this commit extracts the function "testDeviceBindingConditions"
into a new file: binding_conditions_test.go.
No functional changes are introduced.

Signed-off-by: Tsubasa Watanabe <w.tsubasa@fujitsu.com>

ttsuuubasa commented on test/integration/dra/dra_test.go:
Sure, I will do.

ttsuuubasa committed:
scheduler: KEP-5007 add integration tests

- Rescheduling on binding failure using taints or device removal
- Triggering binding timeout when BindingConditions remain unmet
- Recovery to a device without BindingConditions after binding timeout

Signed-off-by: Tsubasa Watanabe <w.tsubasa@fujitsu.com>

ttsuuubasa commented:
The `pull-kubernetes-linter-hints` test appears to be failing due to the use of `gomega.BeTrue` within our tests.
It seems that switching to `gomega.BeTrueBecause` could resolve the issue, so I’m happy to make the fix if required.

pohly commented on test/integration/dra/binding_conditions_test.go:
```diff
@@ -17,30 +17,46 @@ limitations under the License.
+	failureCondition   = "failed"
+)
+
+func testDeviceBindingConditions(tCtx ktesting.TContext, enabled bool) {
```
This function seems to assume that device taints are enabled. That's okay, but please document it.

pohly commented on test/integration/dra/binding_conditions_test.go:
```diff
@@ -230,3 +229,427 @@ func testDeviceBindingConditions(tCtx ktesting.TContext, enabled bool) {
+			}},
+		},
+	}
+	_ = createSlice(tCtx, slice)
```
Drop `_ =`. Perhaps do search/replace, I might have missed some places.

pohly reviewed (approved):
/approve
/lgtm
/milestone v1.35
/hold

We should include this in v1.35, but let's try to do the enhancements that I am suggesting.

pohly commented on test/integration/dra/binding_conditions_test.go:
```diff
@@ -230,3 +229,427 @@ func testDeviceBindingConditions(tCtx ktesting.TContext, enabled bool) {
+
+	// Ensure the ResourceSlice has been created before the binding timeout occurs.
+	sliceCreated := false
+	ktesting.Eventually(tCtx, func(tCtx ktesting.TContext) error {
```
Yes. `ktesting.Eventually` uses gomega.Eventually, so you can use [`gomega.StopTrying`](https://pkg.go.dev/github.com/onsi/gomega#pkg-variables).

I see that the docs only mentioned `gomega.TryAgainAfter` (https://github.com/kubernetes/kubernetes/blob/f38a61181cd4f73a22c688e4d5c0c73d707f6acb/test/utils/ktesting/assert.go#L176-L187) which I need to fix, but the basic code is the same.

pohly commented on test/integration/dra/binding_conditions_test.go:
```diff
@@ -230,3 +220,392 @@ func testDeviceBindingConditions(tCtx ktesting.TContext, enabled bool) {
+		tCtx.ExpectNoError(err, "get pod")
+		return pod
+	}).WithTimeout(wantTO).WithPolling(200 * time.Millisecond).Should(gomega.SatisfyAll(
+		gomega.WithTransform(func(p *v1.Pod) string {
```
Nit: `gomega.HaveField("Spec.NodeName", gomega.BeEmpty())` is shorter.

pohly reviewed (commented):
Some more nits, but already looking quite good.

pohly commented:
Your last update dropped the condition check:
https://github.com/kubernetes/kubernetes/compare/b085ea7c954f2836310fbaaf73f64b64b8443538..17b53eb1012e693ea027a6468f8a6bab564067b8

pohly commented on test/integration/dra/binding_conditions_test.go:
```diff
@@ -0,0 +1,609 @@
+		}, gomega.Not(gomega.ContainElement(v1.PodCondition{
+			Type:   v1.PodScheduled,
+			Status: v1.ConditionTrue,
+		}))),
```
My suggestion was to replace WithTransform with HaveField in general, so this also becomes:


```suggestion
		gomega.HaveField("Status.Conditions", gomega.Not(gomega.ContainElement(v1.PodCondition{
			Type:   v1.PodScheduled,
			Status: v1.ConditionTrue,
		}))),
```

pohly reviewed (approved):
/lgtm

pohly commented:
/hold cancel
